File: python version/speed_limit_rule.py
from typing import Optional
from rule import Rule
from car_type import CarType
from radar_observation import RadarObservation
from violation import Violation

class SpeedLimitRule(Rule):
    def __init__(self, car_type: CarType, max_speed: float, fee: float):
        self.car_type = car_type
        self.max_speed = max_speed
        self.fee = fee
        formatted_type = car_type.value.capitalize()
        self.rule_name = f"{formatted_type} Speed Limit Rule"

    def get_rule_name(self) -> str:
        return self.rule_name

    def evaluate(self, observation: RadarObservation) -> Optional[Violation]:
        if observation.car_type == self.car_type and observation.speed > self.max_speed:
            actual_speed = int(observation.speed)
            limit = int(self.max_speed)
            desc = f"speed of {actual_speed} exceeded max allowed {limit}"
            return Violation(self.rule_name, desc, self.fee)
        return None

File: SpeedLimitRule.java
import java.util.Optional;

public class SpeedLimitRule implements Rule {
    private final CarType carType;
    private final double maxSpeed;
    private final double fee;
    private final String ruleName;

    public SpeedLimitRule(CarType carType, double maxSpeed, double fee) {
        this.carType = carType;
        this.maxSpeed = maxSpeed;
        this.fee = fee;
        this.ruleName = carType.name().substring(0, 1).toUpperCase() + carType.name().substring(1).toLowerCase() + " Speed Limit Rule";
    }

    @Override
    public String getRuleName() {
        return ruleName;
    }

    @Override
    public Optional<Violation> evaluate(RadarObservation observation) {
        if (observation.getCarType() == carType && observation.getSpeed() > maxSpeed) {
            int actualSpeed = (int) observation.getSpeed();
            int limit = (int) maxSpeed;
            String desc = "speed of " + actualSpeed + " exceeded max allowed " + limit;
            return Optional.of(new Violation(ruleName, desc, fee));
        }
        return Optional.empty();
    }
}
